crates/ruff/src/rules/pylint/rules/unexpected_special_method_signature.rs

```diff
@@ -0,0 +1,132 @@
+use std::ops::RangeInclusive;
+
+use rustpython_parser::ast::{Arguments, Expr, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::identifier_range;
+use ruff_python_ast::source_code::Locator;
+use ruff_python_semantic::analyze::visibility::is_staticmethod;
+use ruff_python_semantic::scope::ScopeKind;
+
+use crate::checkers::ast::Checker;
+
+#[violation]
+pub struct UnexpectedSpecialMethodSignature {
+    pub method_name: String,
+    pub expected_params: std::ops::RangeInclusive<usize>,
+    pub actual_params: usize,
+}
+
+impl Violation for UnexpectedSpecialMethodSignature {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let verb = if self.actual_params > 1 {
+            "were"
+        } else {
+            "was"
+        };
+        if self.expected_params.end() - self.expected_params.start() > 1 {
+            format!(
+                "The special method '{}' expects between {} and {} param(s), {} {} given",
+                self.method_name,
+                self.expected_params.start(),
+                self.expected_params.end(),
+                self.actual_params,
+                verb,
+            )
+        } else {
+            format!(
+                "The special method '{}' expects {} param(s), {} {} given",
```

Comment:
Nit: consider using either `parameter` or `parameters` based on `self.expected_params.start()`, rather than `param(s)`? It's a little more code (one more case) but we tend to do it elsewhere.

Reply:
Done.

I was using the copy from pylint as a starting point but this is clearer.